src/main/java/fr/openent/lystore/controllers/PurseController.java

```diff
@@ -261,26 +265,27 @@ private static void renderErrorMessage(HttpServerRequest request, Throwable caus
     public void export(final HttpServerRequest request) {
         try {
             Integer idCampaign = Integer.parseInt(request.params().get("id"));
-            purseService.getPursesByCampaignId(idCampaign, new Handler<Either<String, JsonArray>>() {
-                @Override
-                public void handle(Either<String, JsonArray> event) {
-                    if (event.isRight()) {
-                        JsonArray ids = new fr.wseduc.webutils.collections.JsonArray();
-                        JsonObject exportValues = new JsonObject();
-                        JsonArray purses = event.right().getValue();
-                        JsonObject purse;
-                        for (int i = 0; i < purses.size(); i++) {
-                            purse = purses.getJsonObject(i);
-                            exportValues.put(purse.getString("id_structure"),
-                                    Double.parseDouble(purse.getString("amount")));
-                            ids.add(purse.getString("id_structure"));
-                        }
-                        retrieveUAIs(ids, exportValues, request);
-                    } else {
-                        badRequest(request);
-                    }
-                }
-            });
+            Domain domain = new Domain(getHost(request), I18n.acceptLanguage(request));
+            List<Purse> purseArrayList = new ArrayList<>();
+            purseService.getPursesByCampaignId(idCampaign)
+                    .compose(purses -> {
+                        purseArrayList.addAll(purses);
+                        return structureService.getStructureById(
+                                new JsonArray(purses.stream()
+                                        .map(purse -> purse.getStructure().getId())
+                                        .collect(Collectors.toList())));
+                    })
+                    .compose(structures -> {
+                        Map<Structure, Purse> values = new HashMap<>();
+                        structures.forEach(structure -> values.put(structure,
+                                purseArrayList.stream().filter(purse ->
+                                        purse.getStructure().getId().equals(structure.getId())).findFirst().orElse(null)));
+                        return purseService.getExport(values, domain);
+                    })
```

Comment:
comme je ne vois pas le reste du code. Juste pour ne rien louper : est-ce le fait que puuse peut-être null a bien été géré ensuite ?

Reply:
for (Map.Entry<Structure, Purse> entry : values.entrySet()) {
            if(entry.getValue() != null )
                exportString.append(getCSVLine(entry.getKey(), entry.getValue()));
        }